### ISSUE TITLE
Windows: cannot bootstrap / run prebuilt 0.4.3 with latest MSYS and pacman packages

### ISSUE THREAD

**Issue by laszlocsomor** [bazelbuild/bazel#2449]
**TL;DR**: I uninstalled MSYS, installed the latest version with latest pacman packages. Now I cannot run the prebuilt bazel-0.4.3 binary, nor can I build my own. Investigating.

----
More info -- Copied from https://github.com/bazelbuild/bazel/issues/2431#issuecomment-276030212:

I did the following:

1. removed msys
    Attempted this several times. The uninstaller always went to 100% quickly and just sat there. Turns out it was trying to delete about 20G of data that was my `/tmp` and `/var/tmp` directory. I terminated the uninstaller, deleted these by hand, then the uninstaller could successfully finish.

2. downloaded the latest version (20161025), installed it (no issues seen)

3. ran pacman (bazel doesn't require vim, but I do):
    ```
    pacman -Syuu gcc git curl zip unzip zlib-devel vim
    ```
    On the first try it printed something like "warning: close this windows, reopen msys, try running pacman again", didn't look scary. I did so, and now pacman succeeded without any errors.

4. I could successfully start the prebuilt bazel-0.4.3 binary, the auto-config found Python and VS, but it couldn't build bazel from source because it failed while trying to load the `windows_jni.dll`. There was also a "Visual Studio not found" error.
    ```
    INFO: Found 1 target...
    ERROR: C:/tempdir/cleanbazel/bazel/src/main/native/BUILD:75:1: Executing genrule //src/main/native:windows_jni failed: bash.exe failed: error executing command C:/tools/msys64/usr/bin/bash.exe -c ... (remaining 1 argument(s) skipped): com.google.devtools.build.lib.shell.BadExitStatusException: Process exited with status 1.
    Visual Studio not found
    --version-sortThe system cannot find the file specified.
    ERROR: C:/tempdir/cleanbazel/bazel/src/BUILD:92:5: declared output 'src/dummy_darwin_tools/src/tools/xcode/momcwrapper/momcwrapper' was not created by genrule. This is probably because the genrule actually didn't create this output, or because the output was a directory and the genrule was run remotely (note that only the contents of declared file outputs are copied from genrules run remotely).
    ERROR: C:/tempdir/cleanbazel/bazel/src/BUILD:92:5: declared output 'src/dummy_darwin_tools/src/tools/xcode/swiftstdlibtoolwrapper/swiftstdlibtoolwrapper' was not created by genrule. This is probably because the genrule actually didn't create this output, or because the output was a directory and the genrule was run remotely (note that only the contents of declared file outputs are copied from genrules run remotely).
    Target //src:bazel failed to build
    Use --verbose_failures to see the command lines of failed build steps.
    INFO: Elapsed time: 25.441s, Critical Path: 2.28s
    ```

5. I tried running a bazel version I built a few days ago. On the first try it failed with the following message:
   ```
   $ /c/work/bazel/bazel-dev build src:bazel
   Extracting Bazel installation...

   Failed to write zipped file "/c/Users/LASZLO~1/AppData/Local/Temp/_bazel_laszlocsomor/install/2b15357e4568f22bb12b81a18c87b79f.tmp.5684/_embedded_binaries/embedded_tools/third_party/checker_framework_dataflow/java/org/checkerframework/dataflow/cfg/node/StringConcatenateAssignmentNode.java": Invalid or incomplete multibyte or wide character
   ```

   That's actually the wrong error message (because of https://github.com/bazelbuild/bazel/issues/2411), the actual reason is error 3, i.e. the path is too long:

   ```
   $ echo -n '/c/Users/LASZLO~1/AppData/Local/Temp/_bazel_laszlocsomor/install/2b15357e4568f22bb12b81a18c87b79f.tmp.5684/_embedded_binaries/embedded_tools/third_party/checker_framework_dataflow/java/org/checkerframework/dataflow/cfg/node/StringConcatenateAssignmentNode.java' | wc -c
   260
   ```

6. I tried again with that dev version, now with a shorter `--output_user_root`. Failed again while trying to load the JNI dll ~~(I suspect it's linked against a different msys dll, causing trouble)~~ [EDIT: that's wrong, the JNI DLL isn't linked against MSYS] and there's the VS not found error again.
   ```
   INFO: Found 1 target...
   ERROR: C:/tempdir/cleanbazel/bazel/src/main/native/BUILD:75:1: Executing genrule //src/main/native:windows_jni failed: bash.exe failed: error executing command C:/tools/msys64/usr/bin/bash.exe -c ... (remaining 1 argument(s) skipped): com.google.devtools.build.lib.shell.BadExitStatusException: Process exited with status 1.
   Visual Studio not found
   --version-sortThe system cannot find the file specified.
   ERROR: C:/tempdir/cleanbazel/bazel/src/BUILD:92:5: declared output 'src/dummy_darwin_tools/src/objc_tools/plmerge/plmerge_deploy.jar' was not created by genrule. This is probably because the genrule actually didn't create this output, or because the output was a directory and the genrule was run remotely (note that only the contents of declared file outputs are copied from genrules run remotely).
   ERROR: C:/tempdir/cleanbazel/bazel/src/BUILD:92:5: declared output 'src/dummy_darwin_tools/src/objc_tools/bundlemerge/bundlemerge_deploy.jar' was not created by genrule. This is probably because the genrule actually didn't create this output, or because the output was a directory and the genrule was run remotely (note that only the contents of declared file outputs are copied from genrules run remotely).
   ERROR: C:/tempdir/cleanbazel/bazel/src/BUILD:92:5: declared output 'src/dummy_darwin_tools/src/tools/xcode/stdredirect/StdRedirect.dylib' was not created by genrule. This is probably because the genrule actually didn't create this output, or because the output was a directory and the genrule was run remotely (note that only the contents of declared file outputs are copied from genrules run remotely).
   ERROR: C:/tempdir/cleanbazel/bazel/src/BUILD:92:5: declared output 'src/dummy_darwin_tools/src/tools/xcode/xcrunwrapper/xcrunwrapper' was not created by genrule. This is probably because the genrule actually didn't create this output, or because the output was a directory and the genrule was run remotely (note that only the contents of declared file outputs are copied from genrules run remotely).
   ERROR: C:/tempdir/cleanbazel/bazel/src/BUILD:92:5: declared output 'src/dummy_darwin_tools/src/tools/xcode/swiftstdlibtoolwrapper/swiftstdlibtoolwrapper' was not created by genrule. This is probably because the genrule actually didn't create this output, or because the output was a directory and the genrule was run remotely (note that only the contents of declared file outputs are copied from genrules run remotely).
   ERROR: C:/tempdir/cleanbazel/bazel/src/main/cpp/util/BUILD:74:1: output 'src/main/cpp/util/_objs/port/src/main/cpp/util/port.o' was not created.
   Target //src:bazel failed to build
   Use --verbose_failures to see the command lines of failed build steps.
   INFO: Elapsed time: 2.037s, Critical Path: 0.87s
   ```

7. Downloaded the distribution archive to build bazel from source.
   That failed with an error I haven't seen before.
   ```
   $ ./compile.sh
   INFO: You can skip this first step by providing a path to the bazel binary as second argument:
   INFO:    ./compile.sh compile /path/to/bazel
   🍃  Building Bazel from scratchFile not found - *.jar
   File not found - *.jar
   no error prone jar
   ```

I'll come back with more updates when I have them.

**Comment by laszlocsomor** [comment]
Ah, I remember this error now, it's due to a bad `$PATH`; we don't have `/bin` and `/usr/bin` on the path, so `find` becomes Windows' `find.exe` instead of `/usr/bin/find`.

Solution is:
```
export PATH=/bin:/usr/bin:$PATH
```

Now compile.sh can proceed and fails at the proto compilation stage with import errors such as:
```
C:\Users\laszlocsomor\AppData\Local\Temp\bazel_xPg4RFcT\src\com\google\devtools\build\lib\remote\CasServiceGrpc.java:3: error: package io.grpc.stub does not exist
import static io.grpc.stub.ClientCalls.asyncUnaryCall;
                          ^
C:\Users\laszlocsomor\AppData\Local\Temp\bazel_xPg4RFcT\src\com\google\devtools\build\lib\remote\CasServiceGrpc.java:3: error: static import only from classes and interfaces
import static io.grpc.stub.ClientCalls.asyncUnaryCall;
^
C:\Users\laszlocsomor\AppData\Local\Temp\bazel_xPg4RFcT\src\com\google\devtools\build\lib\remote\CasServiceGrpc.java:4: error: package io.grpc.stub does not exist
import static io.grpc.stub.ClientCalls.asyncServerStreamingCall;
                          ^
```
I keep troubleshooting.

**Comment by laszlocsomor** [comment]
> 6. I tried again with that dev version, now with a shorter --output_user_root. Failed again while trying to load the JNI dll (I suspect it's linked against a different msys dll, causing trouble) and there's the VS not found error again.

That's totally wrong, windows_jni.dll isn't linked against msys.

**Comment by laszlocsomor** [comment]
Managed to get `./compile.sh` working again on Windows.

Turns out I couldn't run it from the 0.4.3 distribution package because of a bug we fixed afterwards, see https://github.com/bazelbuild/bazel/commit/5402993a5e9065984a42eca2132ec56ca3aa456f.

I'll update https://bazel.build/versions/master/docs/windows.html

**Comment by petemounce** [comment]
@laszlocsomor any requirement to update bazel's chocolatey package dependency on msys2 as a consequence of this? It's currently [20160719.1 or later](https://github.com/bazelbuild/bazel/blob/master/scripts/packages/chocolatey/bazel.nuspec.template#L79).

**Comment by laszlocsomor** [comment]
@petemounce : The released bazel.exe is linked against the 20160719 version of msys-2.0.dll, it won't work with newer versions. (You can compile Bazel against newer versions though.) In the light of that I think the requirement should be 20160719.1 exactly.

**Comment by dslomov** [comment]
@laszlocsomor ping

**Comment by laszlocsomor** [comment]
This should be fixed except you have to apply a patch to fix a bug in the bootstrapping scripts, see https://bazel.build/versions/master/docs/install.html#known-issues-when-compiling-from-source

**Comment by pcj** [comment]
Documented my experience getting bazel installed on windows as a Gist:

https://gist.github.com/pcj/46553228eb782221a1bacd68b37b3d49
